internal/cli/server/server.go

```diff
@@ -160,60 +166,123 @@ func initContextMiddleware(logger logr.Logger) gin.HandlerFunc {
 	}
 }
 
-// authMiddleware authenticates the user either using the session or if one
-// doesn't exist, it authenticates with basic auth.
-func authMiddleware(ctx *gin.Context) {
-	reqCtx := ctx.Request.Context()
-	logger := requestctx.Logger(reqCtx).WithName("AuthMiddleware")
+var oidcProvider *dex.OIDCProvider
 
-	userMap, err := loadUsersMap(ctx)
-	if err != nil {
-		response.Error(ctx, apierrors.InternalError(err))
-		ctx.Abort()
-		return
+// getOIDCProvider returns a lazy constructed OIDC provider
+func getOIDCProvider(ctx context.Context) (*dex.OIDCProvider, error) {
+	if oidcProvider != nil {
+		return oidcProvider, nil
 	}
 
-	if len(userMap) == 0 {
-		response.Error(ctx, apierrors.NewAPIError("no user found", http.StatusUnauthorized))
-		ctx.Abort()
-		return
+	mainDomain, err := domain.MainDomain(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting main domain")
 	}
 
-	logger.V(1).Info("Basic auth authentication")
+	issuer := fmt.Sprintf("https://auth.%s", mainDomain)
+	oidcProvider, err = dex.NewOIDCProvider(ctx, issuer, "epinio-cli")
```

Comment:
We are not the `epinio-cli` here, are we? Shouldn't we use an authenticated client whenever possible? We do create a secret for an authenticated client here: https://github.com/epinio/helm-charts/pull/252/files#diff-cd53f64fab50a25f4a05f17fea4f64e7648a1522d85c48decd504d905b6989e8R7

This function is always uses `OutOfBrowserURN` for redirections. I guess it doesn't matter because the server only uses the provider for token validation and not the full login flow (so no redirections happen). Maybe it's fine like this then?

Reply:
or we issue the token with the authenticated client in the audience too?

Reply:
Not sure what you mean. The token was issued for the `epinio-cli` client, so the server is checking that the token has the `epinio-cli` audience in it.

Reply:
https://dexidp.io/docs/custom-scopes-claims-clients/#scopes

```
audience:server:client_id:( client-id ) Dynamic scope indicating that the ID token should be issued on behalf of another client. See the “Cross-client trust and authorized party” section below.
```

Reply:
The example that they do is the possibility to generate token for another client. i.e. you could have a web page in the Epinio UI that can generate a token, that will have the `epinio-cli` audience, even though it was issued by the `epinio-ui`.
But we still need the clientID, and check that the audience is matching what we expect to match.
So I'm not sure how this could help.

Reply:
The `epinio-cli` auth client is generating the tokens but the server auth client (the one we wanted to introduce, the one that has credentials) is the one that checks the validity of the token. Thus, the server auth client is also in the "audience". I think the original request to generate the token, should include the server auth client in the audience so that it can later check the token's validity. At least, that's how I understand this.

Reply:
I'm still not convinced. At the moment we have only one client, that is the `epinio-cli`. The server is not acting as a client, because is not requesting tokens, but it just verifies them.

The audience in the token `MUST contain the OAuth 2.0 client_id of the Relying Party as an audience value. It MAY also contain identifiers for other audiences.` (https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

As far as I understood looking for some other [explanations](https://stackoverflow.com/questions/28418360/jwt-json-web-token-audience-aud-versus-client-id-whats-the-difference) it should be used to identify the resources that the token has access to (i.e: api.epinio.io, and not api3.epinio.io).

Looking at the Dex [doc](https://dexidp.io/docs/custom-scopes-claims-clients/#cross-client-trust-and-authorized-party) they say
> The web app can then use the following scope to request an ID token that’s issued for the command line tool.
> `audience:server:client_id:cli-app`

so in our case we should add a new staticClient (or use/modify the `epinio-ui`?), altough if it's not really a client, and modify the client to ask for an `epinio-server` audience?

```yaml
staticClients:
- id: epinio-server
    secret: {{ $clientSecret }}
    name: 'Epinio Web UI'
    redirectURIs:
    - "https://epinio.{{ .Values.global.domain }}/dex/callback"
    trustedPeers:
    - epinio-cli

- id: epinio-cli
    public: true
    name: 'Epinio cli'
```
so modifying the cli part that asks for the token: https://github.com/epinio/epinio/blob/adc72decd12a82ee717cefe96535e9084dc9edb1/internal/cli/usercmd/login_oidc.go#L53-L56

adding this scope:

```go
oidcProvider.Config.Scopes = append(oidcProvider.Config.Scopes, "audience:server:client_id:epinio-server")
```

But this would simply mean to "hidden" the cli client, and it doesn't seems like the ideal use case.